spec/01-unit/01-db/12-missing-migrations_spec.lua

```diff
@@ -0,0 +1,152 @@
+local pl_dir    = require("pl.dir")
+local pl_path   = require("pl.path")
+
+local MIGRATIONS = {
+    {
+        dir = "kong/db/migrations/core",
+        index = "kong/db/migrations/core/init.lua",
+        skip = {
+            -- FIXME: KAG-6487
+            ["kong/db/migrations/core/025_390_to_3100.lua"] = true,
+        },
+    }
+}
+local PLUGIN_DIRS = {
+    "kong/plugins",
+}
+
+
+--[[
+    input: {
+        "a",
+        "b",
+        "c",
+    }
+
+    output: {
+        a = true,
+        b = true,
+        c = true,
+    }
+--]]
+local function array2map(array)
+    local map = {}
+    for _, value in ipairs(array) do
+        assert(type(value) == "string")
+        map[value] = true
+    end
+    return map
+end
+
+
+--[[
+    input: {
+        a = <anything A>,
+        b = <anything B>,
+        c = <anything C>,
+    }
+
+    output: {
+        <anything A>,
+        <anything B>,
+        <anything C>,
+    }
+--]]
+local function flatten(map)
+    local array = {}
+    for _, v in pairs(map) do
+        table.insert(array, v)
+    end
+    return array
+end
+
+--[[
+    This function will fail for the following cases:
+    1. An migration file is exists but not listed in the `init.lua` file.
+    2. An migration file is marked as skipped but still listed in the `init.lua` file.
+--]]
+local function assert_no_missing_migrations(migrations)
+    for _, migration in ipairs(migrations) do
+        --[[
+            {
+                "010_210_to_211" = true,
+                ...
+            }
+        --]]
+        local indexed_migration = array2map(loadfile(migration.index)())
+        local files = pl_dir.getfiles(migration.dir)
+        for _, file in ipairs(files) do
+            if file == migration.index then
+                goto continue
+            end
+
+            -- @file: kong/db/migrations/core/010_210_to_211.lua
+            -- @basename: 010_210_to_211.lua
+            local basename = pl_path.basename(file)
+            assert.True(basename:sub(-4) == ".lua", "unexpected file " .. basename)
```

Comment:
I think we could just ignore non-lua files. For something like `README.md` which could be added later.